http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/internal/net/InputShutdownCheckingSslSocket.java

```diff
@@ -28,22 +34,114 @@
  */
 @SdkInternalApi
 public final class InputShutdownCheckingSslSocket extends DelegateSslSocket {
+    private static final Logger LOG = Logger.loggerFor(InputShutdownCheckingSslSocket.class);
+    private final SSLSocket sslSocket;
+    private final Socket layered;
+    private final ByteBuffer readBuffer = ByteBuffer.allocate(1024);
 
-    public InputShutdownCheckingSslSocket(SSLSocket sock) {
-        super(sock);
+    public InputShutdownCheckingSslSocket(SSLSocket sslSocket, Socket layered) {
+        super(sslSocket);
+        this.sslSocket = sslSocket;
+        this.layered = layered;
     }
 
     @Override
     public OutputStream getOutputStream() throws IOException {
-        return new InputShutdownCheckingOutputStream(sock.getOutputStream(), sock);
+        return new InputShutdownCheckingOutputStream(sslSocket.getOutputStream());
     }
 
-    private static class InputShutdownCheckingOutputStream extends FilterOutputStream {
-        private final SSLSocket sock;
+    @Override
+    public InputStream getInputStream() throws IOException {
+        return new BufferedInputStream(sslSocket.getInputStream());
+    }
+
+    // Attempt 1-byte read from the SSLSocket so that the SSLSocket reads the TLS record if there is one pending.
+    private void bufferSslSocketRead() throws IOException {
+        LOG.debug(() -> "Buffering SSLSocket data to consume pending TLS records that may contain a close_notify");
+        synchronized (readBuffer) {
+            if (!readBuffer.hasRemaining() || !dataAvailableOnLayeredSocket()) {
+                return;
+            }
+
+            InputStream is = sslSocket.getInputStream();
+            int originalSoTimeout = sslSocket.getSoTimeout();
+            try {
+                // Maximum time to block for the least amount of time possible (1ms)
+                sslSocket.setSoTimeout(1);
+                int read = is.read();
+                if (read != -1) {
+                    readBuffer.put((byte) read);
+                }
+            } catch (SocketTimeoutException to) {
+                LOG.debug(() -> "Timeout doing single byte read on SSLSocket. Most likely due to incomplete record", to);
+            } finally {
+                sslSocket.setSoTimeout(originalSoTimeout);
+            }
+        }
+    }
+
+    private class BufferedInputStream extends FilterInputStream {
+        private final byte[] oneByte = new byte[1];
```

Comment:
`LookaheadInputStream` from `utils` does this same thing - allow peeking one byte in advance. Not sure if it's possible to reuse that, but would be nice to avoid duplication if possible.